Products/ZenModel/migrate/tests/common.py

```diff
@@ -0,0 +1,136 @@
+from servicemigration import context, service
+
+
+def fakeContextFromFile(jsonfile):
+    jsonfile = os.path.join(os.path.dirname(__file__), jsonfile)
+    class FakeServiceContext(context.ServiceContext):
+
+        def __init__(self, filename=None):
+            self.services = []
+            self.changes = {}
+            for datum in json.loads(open(jsonfile, 'r').read()):
+                self.services.append(service.deserialize(datum))
+            self.version = self.services[0]._Service__data["Version"]
+
+        def commit(self, filename=None):
+            addedServices = []
+            modifiedServices = []
+            for svc in self.services:
+                serial = service.serialize(svc)
+                serial["Version"] = self.version
+                if serial["ID"] == "new-service":
+                    addedServices.append(serial)
+                else:
+                    modifiedServices.append(serial)
+            self.changes = {
+                "ServiceID": "FakeServiceContext",
+                "Modified": modifiedServices,
+                "Added": addedServices,
+                "Deploy": None,
+            }
+
+        def servicedef(self):
+            return [service.serialize(s) for s in self.services]
+                    
+    return FakeServiceContext()
+
+
+def compare_lists(this, that, path=None):
+    path = path or []
+    if len(this) != len(that):
+        return False, path
+    for i, (a, b) in enumerate(zip(this, that)):
+        if isinstance(a, list):
+            r, p = compare_lists(a, b, path + [i])
+            if not r:
+                return False, p
+        elif isinstance(a, dict):
+            r, p = compare_dicts(a, b, path + [i])
+            if not r:
+                return False, p
+        else:
+            if a != b:
+                return False, p
+    return True, None
+
+
+def compare_dicts(this, that, path=None):
+    path = path or []
+    if len(this.keys()) != len(that.keys()):
+        return False, path
+    for k in this.keys():
+        a, b = this[k], that[k]
+        if isinstance(a, list):
+            r, p = compare_lists(a, b, path + [k])
+            if not r:
+                return False, p
+        elif isinstance(a, dict):
+            r, p = compare_dicts(a, b, path + [k])
+            if not r:
+                return False, p
+        else:
+            if a != b:
+                return False, path
+    return True, None
+
+
+class ServiceMigrationTestCase(object):
+    """
+    Superclass of service migration tests.
+    Supply the migration module and class, the input servicedef file,
+    and the output servicedef file.
+    """
+    initial_servicedef = ''
+    expected_servicedef = ''
+    migration_module_name = ''
+    migration_class_name = ''
+
+    def setUp(self):
+        pass
+
+    def tearDown(self):
+        pass
+
+    def test_cutover_correctness(self):
+        context = fakeContextFromFile(self.initial_servicedef)
```

Comment:
A little copy/paste code here which isn't so bad, but I think the differences btwn the two tests would be more clear if they each shared a common method. Something like
```
def test_cutover_correctness(self):
    verify_cutover(self.initial_servicedef, self.expected_servicedef)

 def test_cutover_idempotence(self):
     verify_cutover(self.expected_servicedef, self.expected_servicedef)

```